### PR TITLE
8283643: [AIX, testbug] MachCodeFramesInErrorFile test fails to find 'Native frames' text

### PR DESCRIPTION
**Pull request by backwaterred**
I observed a failure of MachCodeFramesInErrorFile.java on AIX. The raised error complains that 'Native frames: ' is missing. To fix the failure, I made two changes to the test that allowed it to pass on AIX.

The first change made was to the test's `extractFrames` method. AIX overrides `os::platform_print_native_stack` with stack printing specific to AIX/Power. This code does not produce the 'Native frames: ' text that `vmError::print_native_stack` does, so the test was modified to expect the AIX output when run on AIX. [Edit: The changes made to `extractFrames` are now minimal. I made the change to `AixNativeCallstack::print_callstack_for_context` instead.]

A second change was made to the address passed to the `crashInNative1` method. AIX does not consider low-address memory to be protected as other platforms do. In fact, "the first 256 bytes are reserved for software use" according to the Power ISA. Accordingly, the read address for PPC was changed to -1 (0xFF..FF) to produce the expected failure result.

### Testing

Tier1 tests complete as expected on AIX/Power.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8283643](https://bugs.openjdk.java.net/browse/JDK-8283643): [AIX, testbug] MachCodeFramesInErrorFile test fails to find 'Native frames' text

### Reviewers
 * [Doug Simon](https://openjdk.java.net/census#dnsimon) (@dougxc - Committer) ⚠️ Review applies to d0a05457a55866fa7df5cbca2501721bf50b3e9b
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to d0a05457a55866fa7df5cbca2501721bf50b3e9b

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8094/head:pull/8094` \
`$ git checkout pull/8094`

Update a local copy of the PR: \
`$ git checkout pull/8094` \
`$ git pull https://git.openjdk.java.net/jdk pull/8094/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8094`

View PR using the GUI difftool: \
`$ git pr show -t 8094`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8094.diff">https://git.openjdk.java.net/jdk/pull/8094.diff</a>

</details>

### CODE REVIEW

**Comment by backwaterred** [comment]
Is anyone able to spare the time to give this PR a quick review? The changes are minimal, and it fixes a failing AIX test :-)

File: test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java [hunk 1]
```diff
@@ -71,7 +71,8 @@ public static void main(String[] args) throws Exception {
                 crashInJava1(10);
             } else {
                 assert args[0].equals("crashInVM");
-                crashInNative1(10);
+                // AIX does not prohibit low address reads
+                crashInNative1( Platform.isPPC() ? -1 : 10 );
```
Comment (TheRealMDoerr) [inline test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java:75]: Spaces next to braces are uncommon.

File: test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java [hunk 2]
```diff
@@ -164,11 +165,16 @@ private static void run(boolean crashInJava) throws Exception {
...
      * and adds them to {@code frames}.
      */
     private static void extractFrames(String hsErr, Set<String> frames, boolean nativeStack) {
-        String marker = (nativeStack ? "Native" : "Java") + " frames: ";
+        String marker;
+        if (Platform.isAix()) {
+            marker = nativeStack ? "------ current frame:" : "Java frames: ";
+        } else {
+            marker = (nativeStack ? "Native" : "Java") + " frames: ";
+        }
```
Comment (TheRealMDoerr) [inline test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java:177]: Yeah, we have special stack trace code on AIX. Ok.

File: test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java [hunk 3]
```diff
@@ -180,6 +186,6 @@ private static void extractFrames(String hsErr, Set<String> frames, boolean nati
                 frames.add(line);
             }
         }
-        throw new RuntimeException("\"" + marker + "\" line missing in hs_err_pid file:\n" + hsErr);
+        throw new RuntimeException("\"" + marker + "\" line missing in hs_err_pid file:\n");
```
Comment (TheRealMDoerr) [inline test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java:189]: Please explain!

**Comment by TheRealMDoerr** [comment]
Looks basically good to me, but please take a look at my comments.

**Comment by tstuefe** [comment]
Hi Tyler,

in the JVM we have `VMError::segfault_address` for this reason. I'd prefer if we reused that (publishing it via Whitebox method) instead of polluting the shared tests with PPC specifics. We could reuse this whitebox method in other tests, actually, without running into the NULL-is-readable issue again.

(To see how whitebox methods are used, see the classes in sun.hotspot.WhiteBox and their corresponding endpoints in hotspot/share/prims/whitebox.hpp. Its all pretty straightforward).

Another, maybe simpler option would be just to write instead of read in `crashInNative3`. crashInJava3 uses write too (putLong). But it looks deliberate, so you'd have to look at the original RFR for this test.

Cheers, Thomas

**Comment by backwaterred** [comment]
Hi Thoams,

> in the JVM we have `VMError::segfault_address` for this reason. I'd prefer if we reused that (publishing it via Whitebox method) instead of polluting the shared tests with PPC specifics. We could reuse this whitebox method in other tests, actually, without running into the NULL-is-readable issue again.
>
> (To see how whitebox methods are used, see the classes in sun.hotspot.WhiteBox and their corresponding "WB_..." endpoints in hotspot/share/prims/whitebox.cpp. Its all pretty straightforward).

I like this idea. It's not really reasonable to expect that everyone will keep track of all the little nuances between the different systems, and it's easier if we solve the problem one way and use it in multiple places. I will take a look at this.

> Another, maybe simpler option would be just to write instead of read in `crashInNative3`. crashInJava3 uses write too (putLong). But it looks deliberate, so you'd have to look at the original RFR for this test.

Tried that ;-) Your intuition is good; the crash produces slightly different output than the test expects.

File: test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java (hunk 3 above)
Comment (backwaterred) [reply test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java:189]: OK!
This was an odd one; I was getting errors that complained about the length of the message in the exception. It looked like 2^16 was the max number of characters that could be included in a RuntimeException. I haven't found the upper limit documented anywhere, but I would get unexpected failures with it included in the message. The text prints to stdout, so I didn't suffer by removing it.

File: test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java (hunk 2 above)
Comment (backwaterred) [reply test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java:177]: :+1:

File: test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java (hunk 1 above)
Comment (backwaterred) [reply test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java:75]: I will drop them.

**Comment by tstuefe** [comment]
BTW, feel free to tweak the native stack printing on os_aix.cpp to print "Native" etc like the other variants do. It is just cosmetics, and the less AIX specifics in shared code the better.

Thanks for taking care of this!

File: test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java (hunk 3 above)
Comment (TheRealMDoerr) [reply test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java:185]: Maybe just print the last line?

Comment (backwaterred) [reply test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java]: I did a bit more testing to jog my memory (it has been a little while since I made these changes). You were right to be concerned about the removal; the error text is not printed automatically.

To avoid having an exception message be too long, I printed the String to std[out|err] using the usual System.[out|err].println methods.

Good catch :slightly_smiling_face:

**Comment by backwaterred** [comment]
> ...feel free to tweak the native stack printing on os_aix.cpp to print "Native" etc like the other variants do. It is just cosmetics, and the less AIX specifics in shared code the better.

This worked well! I incorporated this change.

Regarding publishing `VMError::segfault_address` via a Whitebox method. I think this is meaty enough to warrant it's own issue and PR. I'll add that to my todo list and tag you when it's ready for review. In that PR I will scan through the project for tests where the change can be used and make changes as appropriate.

My AIX testing is again passing. Converting this back to and open PR.

File: test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java [hunk 4]
```diff
@@ -127,9 +128,9 @@ private static void run(boolean crashInJava) throws Exception {
         // Extract hs_err_pid file.
         String hs_err_file = output.firstMatch("# *(\\S*hs_err_pid\\d+\\.log)", 1);
         if (hs_err_file == null) {
-            throw new RuntimeException("Did not find hs_err_pid file in output.\n" +
-                                       "stderr:\n" + output.getStderr() + "\n" +
-                                       "stdout:\n" + output.getStdout());
+            System.out.println("stdout:\n" + output.getStdout());
+            System.err.println("stderr:\n" + output.getStderr());
+            throw new RuntimeException("Did not find hs_err_pid file in output");
```
Comment (dougxc) [inline test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java:133]: Why this change (and a similar one at the end of the file)? This separates the reporting of a failed condition (i.e., could not find expected patten in output) from the context (i.e. contents of stderr and stdout) that assists in diagnosing why the condition may have failed.

Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java:133]: Also the test should really take advantage of output.shouldContain to check for the hs_err line, which in turn will report stdout and stderr on failure, and throw the exception.

File: test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java [hunk 5]
```diff
@@ -180,6 +182,7 @@ private static void extractFrames(String hsErr, Set<String> frames, boolean nati
                 frames.add(line);
             }
         }
-        throw new RuntimeException("\"" + marker + "\" line missing in hs_err_pid file:\n" + hsErr);
+        System.err.println(hsErr);
```
Comment (backwaterred) [inline test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java:185]: It's a popular (and reasonable) [question](https://github.com/openjdk/jdk/pull/8094#discussion_r852787701) as to why I made the change to the exception reporting. This is the error that resulted from a too-long exception message.

```
java.io.UTFDataFormatException: encoded string (`main' t....11 END.) too long: 116222 bytes
        at java.base/java.io.DataOutputStream.writeUTF(DataOutputStream.java:369)
        at java.base/java.io.DataOutputStream.writeUTF(DataOutputStream.java:333)
        at com.sun.javatest.regtest.agent.AgentServer.writeStatus(AgentServer.java:351)
        at com.sun.javatest.regtest.agent.AgentServer.doMain(AgentServer.java:307)
        at com.sun.javatest.regtest.agent.AgentServer.run(AgentServer.java:233)
        at com.sun.javatest.regtest.agent.AgentServer.main(AgentServer.java:70)
```

This error hampered my ability to debug this test. It's not very clear from the message as to where it's coming from, and it occurs sporadically depending on the length of the message produced. I eventually traced it to the line: `throw new RuntimeException("\"" + marker + "\" line missing in hs_err_pid file:\n" + hsErr);`. To trigger the error for this comment, I changed `hsErr` to `hsErr + hsErr`.

This issue may be platform specific, or the issue just isn't encountered very often because it's uncommon to add > 64KiB Strings to an Exception. Either way, making this change made the test failures much easier to diagnose.

File: test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java (hunk 4 above)
Comment (backwaterred) [reply test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java:133]: @dougxc, I added [an explanation](https://github.com/openjdk/jdk/pull/8094/files#r855362995) to the similar change at the end of the file. That was the one I discovered first. Here I made a similar change to prevent the same issue from occurring.

@dholmes-ora, I don't think output.shouldMatch is the best choice here because it involves doing the same operation twice. I believe I would end up looking for the pattern and throwing an error with shouldMatch, then looking for the pattern again and returning the String with firstMatch. I do like your idea of using more standard code however. I would like to replace the calls to println with `output.reportDiagnosticSummary()` if that is a reasonable compromise to your suggestion?

Comment (backwaterred) [reply test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java:133]: @dougxc, I added [an explanation](https://github.com/openjdk/jdk/pull/8094/files#r855362995) to the similar change at the end of the file. That was the one I discovered first. Here I made a similar change to prevent the same issue from occurring.

@dholmes-ora, I don't think output.shouldMatch is the best choice here because it involves doing the same operation twice. I believe I would end up looking for the pattern and throwing an error with shouldMatch, then looking for the pattern again and returning the String with firstMatch. I do like your idea of using more standard code however. I would like to replace the calls to println with `output.reportDiagnosticSummary()` if that is a reasonable compromise to your suggestion?

Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java:132]: It is a test so I'm not overly concerned about scanning the output twice - though I agree it is a bit irksome to have to do that (new RFE: add `firstMatchOrFail`?). Using `reportDiagnosticSummary` is a reasonable compromise. Thanks.

**Comment by backwaterred** [comment]
I believe all requested changes are complete, but I don't have any reviews. Are there any changes you'd like to see before merge @TheRealMDoerr, @dougxc, or @dholmes-ora?

**Review by dougxc — APPROVED** [review]

File: test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java [hunk 6]
```diff
@@ -71,7 +71,8 @@ public static void main(String[] args) throws Exception {
                 crashInJava1(10);
             } else {
                 assert args[0].equals("crashInVM");
-                crashInNative1(10);
+                // AIX does not prohibit low address reads
+                crashInNative1(Platform.isPPC() ? -1 : 10);
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java:75]: Is it AIX or PPC or AIX/PPC that needs the change? The code should match the comment.

**Review by dholmes-ora — APPROVED** [review]
Change seems reasonable but please fix the issue below before integrating.

Thanks,
David

File: test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java (hunk 6 above)
Comment (backwaterred) [reply test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java:75]: I remember changing this now. It's actually PPC. The Power ISA specifies that the first 256 bytes are 'reserved for software use'.

I'll change the comment.

**Comment by backwaterred** [comment]
Thanks very much for your reviews and comments.
/integrate

File: test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java (hunk 6 above)
Comment (tstuefe) [reply test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java]: Really? I was always assuming the linux kernel handles this differently.

Comment (TheRealMDoerr) [reply test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java]: I think the mentioned Power ISA chapter addresses real addresses, not the user mode virtual address space. We always get SIGSEGV when accessing the first page on linux.

Comment (tstuefe) [reply test/hotspot/jtreg/runtime/ErrorHandling/MachCodeFramesInErrorFile.java]: Thanks for clarifying!
